### PR TITLE
Lazily import `torch/pydantic` in `json` module, speedup `from monty.json import` by 10x

### PR DESCRIPTION
**Pull request by DanielYang59**
### Summary

- Lazily import `torch/pydantic` for `monty.json`, thanks for @janosh's profiling in https://github.com/janosh/pymatviz/issues/209#issuecomment-2423174822, would significantly speed up `from monty.json import MSONable/MontyEncoder/MontyDecoder/...`

---

## Performance comparison

- Following benchmark done with **optional dependencies installed** (otherwise import would be skipped), averaged across 10 runs
- Python 3.12.7, MacOS 15.0.1 with M3 chip
- Profile with: `python -X importtime -c  "from monty.json import MSONable" 2> monty.log && tuna monty.log`

<details>
<summary> Benchmark Script (by GPT) </summary>

```python
import time
import subprocess


def measure_import_time(command: str, repeats: int):
    total_time = 0

    for _ in range(repeats):
        start_time = time.time()

        # Run the Python command in a new process
        subprocess.run(["python", "-c", command], stdout=subprocess.PIPE, stderr=subprocess.PIPE)

        end_time = time.time()
        total_time += (end_time - start_time)

    avg_time_ms = (total_time / repeats) * 1000
    print(f"Average import time for '{command}': {avg_time_ms:.3f} ms")

def main():
    commands = [
        "from monty.json import MSONable",
        "from monty.json import MontyEncoder",
        "from monty.json import MontyDecoder",
    ]

    for command in commands:
        measure_import_time(command, repeats=10)

if __name__ == "__main__":
    main()

```

</details>

### Import time from master branch
```
Average import time for 'from monty.json import MSONable': 924.585 ms
Average import time for 'from monty.json import MontyEncoder': 735.797 ms
Average import time for 'from monty.json import MontyDecoder': 720.402 ms
```

Import time traced for `from monty.json import MSONable`:
image

### Import time from tip of this PR
```
Average import time for 'from monty.json import MSONable': 109.053 ms
Average import time for 'from monty.json import MontyEncoder': 111.683 ms
Average import time for 'from monty.json import MontyDecoder': 103.302 ms
```

image

---

### After lazily import `torch`

```
Average import time for 'from monty.json import MSONable': 122.606 ms
Average import time for 'from monty.json import MontyEncoder': 123.456 ms
Average import time for 'from monty.json import MontyDecoder': 117.621 ms
```

image

---

### After lazily import `numpy` as well (reverted, as use of np is very common)

```
Average import time for 'from monty.json import MSONable': 58.586 ms
Average import time for 'from monty.json import MontyEncoder': 59.323 ms
Average import time for 'from monty.json import MontyDecoder': 58.300 ms
```

image

---

### After lazily import `pydantic` too (limited improvement from here)

```
Average import time for 'from monty.json import MSONable': 54.003 ms
Average import time for 'from monty.json import MontyEncoder': 53.038 ms
Average import time for 'from monty.json import MontyDecoder': 52.875 ms
```

image

---

### After lazily import `ruamel.yaml` (reverted, no obvious improvement)

```
Average import time for 'from monty.json import MSONable': 52.769 ms
Average import time for 'from monty.json import MontyEncoder': 52.191 ms
Average import time for 'from monty.json import MontyDecoder': 60.379 ms
```

image

## Summary by CodeRabbit

- **New Features**
	- Enhanced type safety and error handling in JSON processing.
- **Chores**
	- Updated GitHub Actions workflow for linting to use the latest Python setup action.
	- Modified pre-commit hook configuration to run codespell during the pre-commit phase.
	- Updated testing workflow to include Python 3.12 in the testing environment.

### CODE REVIEW

File: src/monty/json.py [hunk 1]
```diff
@@ -21,47 +21,24 @@
...
-    import torch
-except ImportError:
-    torch = None
-
 __version__ = "3.0.0"
 
 
 def _load_redirect(redirect_file):
     try:
+        from ruamel.yaml import YAML
```
Comment (DanielYang59) [inline src/monty/json.py:40]: This PR doesn't change the overall workflow of `monty`, just a head up here, if `ruamel.yaml` is not install, trying to import `MSONable` would trigger `ImportError`.

Also note `monty` as a standalone package doesn't list any dependencies: https://github.com/materialsvirtuallab/monty/blob/1270c7b230b43cdbc4f5f184ac645ece31d49b61/pyproject.toml#L21-L23

File: src/monty/json.py [hunk 2]
```diff
@@ -586,8 +566,8 @@ def default(self, o) -> dict:
         if isinstance(o, Path):
             return {"@module": "pathlib", "@class": "Path", "string": str(o)}
 
-        if torch is not None and isinstance(o, torch.Tensor):
-            # Support for Pytorch Tensors.
+        # Support for Pytorch Tensors
+        if _check_type(o, "torch.Tensor"):
```
Comment (DanielYang59) [inline src/monty/json.py:570]: Should be the correct type string:
```
import torch


tensor = torch.randn(3, 3)
print(type(tensor).mro())
```

Gives: `[<class 'torch.Tensor'>, <class 'torch._C.TensorBase'>, <class 'object'>]`

File: src/monty/json.py [hunk 3]
```diff
@@ -660,7 +644,7 @@ def default(self, o) -> dict:
                 raise AttributeError(e)
 
         try:
-            if pydantic is not None and isinstance(o, pydantic.BaseModel):
+            if _check_type(o, "pydantic.main.BaseModel"):
```
Comment (DanielYang59) [inline src/monty/json.py:647]:
```python
from pydantic import BaseModel


class MyModel(BaseModel):
    name: str

model_instance = MyModel(name="monty")
print(type(model_instance).mro())
```

Gives: `[<class '__main__.MyModel'>, <class 'pydantic.main.BaseModel'>, <class 'object'>]`

**Comment by shyuep** [comment]
I don't think we need to lazy import numpy. Numpy is used everywhere. We need to keep in mind that the import cost will be incurred regardless. I am fine with lazy importing time, pydantic and torch since those are used much less commonly.

**Comment by DanielYang59** [comment]
Cool thanks for the comment! Sure thing I would revert the part for `numpy`, can you please review #714 which would unblock this PR (otherwise nest those import block would make it really hard to see what optional dependency is required)? Really appreciate that

**Comment by shyuep** [comment]
I merged 714

**Comment by DanielYang59** [comment]
Thank you again! Would clean up this PR now

**Comment by DanielYang59** [comment]
@shyuep ready for review, thanks

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -32,7 +32,7 @@ repos:
     rev: v2.3.0
     hooks:
       - id: codespell
-        stages: [commit, commit-msg]
+        stages: [pre-commit, commit-msg]
```
Comment (DanielYang59) [inline .pre-commit-config.yaml:35]:
```
pre-commit run --all-files
[WARNING] hook id `codespell` uses deprecated stage names (commit) which will be removed in a future version.  run: `pre-commit migrate-config` to automatically fix this.
ruff.....................................................................Passed
ruff-format..............................................................Passed
check yaml...............................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
mypy.....................................................................Passed
codespell................................................................Passed
cython-lint..........................................(no files to check)Skipped
double-quote Cython strings..........................(no files to check)Skipped
blacken-docs.............................................................Passed
markdownlint.............................................................Passed
```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -9,7 +9,7 @@ jobs:
       max-parallel: 20
       matrix:
         os: [ubuntu-latest, macos-14, windows-latest]
-        python-version: ["3.9", "3.x"]
```
Comment (DanielYang59) [inline .github/workflows/test.yml:12]: Looks like setup action [sometimes install python 3.12](https://github.com/materialsvirtuallab/monty/actions/runs/11432932326/job/31804141443), and sometimes 3.13, [if 3.13 is installed](https://github.com/materialsvirtuallab/monty/actions/runs/11438818081/job/31821218765?pr=713), would trigger the following error on MacOS:
```
ERROR: Ignored the following versions that require a different python version: 1.21.2 Requires-Python >=3.7,<3.11; 1.21.3 Requires-Python >=3.7,<3.11; 1.21.4 Requires-Python >=3.7,<3.11; 1.21.5 Requires-Python >=3.7,<3.11; 1.21.6 Requires-Python >=3.7,<3.11; 1.26.0 Requires-Python <3.13,>=3.9; 1.26.1 Requires-Python <3.13,>=3.9
ERROR: Could not find a version that satisfies the requirement torch; extra == "json" (from monty[dev,json,multiprocessing,serialization]) (from versions: none)
ERROR: No matching distribution found for torch; extra == "json"
```

**Pull request merged**